File: shellSort/python/shell_sort.py
def shellSort(arr):
    gap = int((len(arr)/2))
    while gap > 0:
        for i in range (gap, len(arr)):
            temp = arr[i]
            j = i
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j-gap]
                j -= gap
            arr[j] = temp
        gap/=2  
        gap = int(gap)


def main():
    arr = [15, 12, 36, 63, 96]
    sorted_arr = shellSort(arr)
    print(sorted_arr)

    
if __name__ == '__main__':
    main()

File: shell_sort/ShellSort.java
import java.util.Arrays;

public class ShellSort {

    // Driver method for testing sort algorithm
    public static void main(String[] args) {
        int[] array = {21, 3, 82, 12, 93, 56, 74, 38};
        System.out.println("Before sort: " + Arrays.toString(array));
        shellSort(array);
        System.out.println("After sort: " + Arrays.toString(array));
    }

    /**
     *
     * Sorts an array by implementing the shell sort algorithm.
     * Shell sort is an in-place comparison sort
     * <p>
     * Average case = depends on the gap<br>
     * Worst case = O(n * log^2 n)<br>
     * Best case = O(n)<br>
     *
     * @param array an unsorted array
     */
    public static void shellSort(int[] array) {
        int inner, outer;
        int temp;

        int h = 1;
        while (h <= array.length / 3) {
            h = h * 3 + 1;
        }
        while (h > 0) {
            for (outer = h; outer < array.length; outer++) {
                temp = array[outer];
                inner = outer;

                while (inner > h - 1 && array[inner - h] >= temp) {
                    array[inner] = array[inner - h];
                    inner -= h;
                }
                array[inner] = temp;
            }
            h = (h - 1) / 3;
        }
    }
}

